### ISSUE TITLE
Issue running virtual rules with multi-element snippets

### ISSUE THREAD

**Issue by nolandanley** [dequelabs/axe-core#3408]
Product: axe-core
Expectation: Running virtual rule on virtual nodes produces findings for nested elements
Actual: Returns "cantTell"
axe-core version: 4.4.1

I’m trying to run virtual rules on code snippets and am seemingly running into issues setting up virtual nodes. I’ve read the [developer guide’s sections on virtual nodes](https://github.com/dequelabs/axe-core/blob/master/doc/developer-guide.md#virtual-nodes) and have not been able to unblock this issue so far.

The following code works perfectly in my local set-up. It raises a violation since the `img` element is missing alt text. When alt text is provided, no violation is raised.

```
const DOM = new JSDOM('<img src="/image.png">');
const node = DOM.window.document.querySelector('img');
const virtualNode = new axe['VirtualNode'](node);
axe['runVirtualRule']('image-alt', virtualNode);
```

But when I want to run a rule that requires multiple elements, I cannot figure out how I need to set things up for them to raise violations. Consider the following code:

```
const DOM = new JSDOM('<div role="text"><a href="/site.html">Link</a></div>');
const parentNode = DOM.window.document.querySelector('div');
const node = DOM.window.document.querySelector('a');
const virtualParentNode = new axe['VirtualNode'](parentNode);
const virtualNode = new axe['VirtualNode'](node, virtualParentNode);
virtualParentNode.children.push(virtualNode);
axe['runVirtualRule']('aria-text', virtualParentNode);
```

This returns “cantTell”. I’ve verified with a [print-out of the `virtualParentNode`](https://pastebin.com/W9B4U6fe) that the nodes are being set up as I expect. If I don’t push the child node into the parent node’s `children` array, it will “pass” (since it’s not aware of any violative nested elements). I’m unsure if I’m not setting up the virtual nodes with everything they need or if there needs to be some other relationship established between them.

I’m aware there are functions like [`createGrid`](https://github.com/dequelabs/axe-core/blob/725a20c91b9006e64009059f0ab9d1a0098d29df/lib/commons/dom/get-rect-stack.js#L12) that seem to manage setting up the virtual nodes and their relationships, but this seems to utilize internal values like the window (which I don’t have since I’m not running this in a browser).

I’m unsure if the next step to get the desired “failed” result is to
- fix a minor issue in the virtual node set-up,
- set up the axe environment in a way that “fakes” the window with the provided code snippet, or
- do something else entirely.

Any advice is greatly appreciated! Maybe adding something to the existing documentation would help others like me down the road.

Constraints:
- I’m not running axe in a browser, so this is all being done with Virtual Nodes that I’m creating from manual JSDOM elements. I’m using [Node.js](https://nodejs.org/en/) for this.
- I’m writing this in TypeScript. Not everything used in the examples above are defined/exposed in the [.d.ts file](https://github.com/dequelabs/axe-core/blob/develop/axe.d.ts). I’ve added these locally and intend to contribute back where it’s possible and makes sense to do so!
- I’m only running on code snippets, not full documents. I’m aware this means some rules will not work as expected and I’m only focusing on particular rules that will hold up in smaller contexts (e.g. [aria-text](https://dequeuniversity.com/rules/axe/4.4/aria-text) and [image-alt](https://dequeuniversity.com/rules/axe/4.4/image-alt) mentioned above.)

**Comment by straker** [comment]
Thanks for the issue. From your code, you should be able to run axe directly on the JSDOM element without going through the virtual rule api. We have a [JSDOM example](https://github.com/dequelabs/axe-core/blob/develop/doc%2Fexamples%2Fjsdom%2Ftest%2Fa11y.js) that shows how you can do that. From there you can use the `runOnly` options to `axe.run` to specify a specific rule to run.

**Comment by straker** [comment]
Interesting. Our JSDOM example also runs in a Node environment so I'm not sure why `setTimeout` would not be available? What version of JSDOM? Anything else about your environment that could help me try to reproduce the error on my end?

**Comment by nolandanley** [comment]
You're absolutely right! I dug into the set-up I was using and found something someone had added that changed that function and was able to mitigate that issue. Was just a red herring that I thought it was inherently something to do with Node.js. Thank you for the help!
